### PR TITLE
lns lookup consensus

### PR DESCRIPTION
**Pull request by majestrate**
ask all service nodes we can for each lns lookup and only use the result if they all agree

### CODE REVIEW

**Review by tewinget — CHANGES_REQUESTED** [review]
the result callback has non-obvious and ideally unnecessary side-effects, and is generally cursed.

File: llarp/service/endpoint.cpp [hunk 1]
```diff
@@ -808,12 +808,43 @@ namespace llarp
...
+        if (path->IsReady())
+        {
+          paths.insert(path);
+        }
+      });
+      // not enough paths
+      if (paths.size() < 3)
+      {
+        handler(std::nullopt);
+        return true;
```
Comment (tewinget) [inline llarp/service/endpoint.cpp]: function now always returns true; should change to void

File: llarp/service/lns_tracker.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      std::unordered_set<Address> m_CurrentValues;
+      std::function<void(std::optional<Address>)> m_HandleResult;
+      std::size_t m_ResultsGotten = 0;
+      std::size_t m_ResultsNeeded;
+
+      LookupInfo(std::size_t wantResults, std::function<void(std::optional<Address>)> resultHandler)
+          : m_HandleResult{std::move(resultHandler)}, m_ResultsNeeded{wantResults}
+      {}
+
+      /// return true to remove self
```
Comment (tewinget) [inline llarp/service/lns_tracker.hpp]: that's a very un-intuitive use of a boolean return.

perhaps add a function "Done" or similar, then above when using do

```
if (itr = find):
    itr->HandleOneResult;
    if itr->Done():
        erase(itr)
```

File: llarp/service/endpoint.cpp [hunk 3]
```diff
@@ -808,12 +808,43 @@ namespace llarp
...
+        handler(result);
+      };
+
+      auto resultHandler =
+          m_state->lnsTracker.MakeResultHandler(name, paths.size(), maybeInvalidateCache);
+
+      for (const auto& path : paths)
+      {
+        LogInfo(Name(), " lookup ", name, " from ", path->Endpoint());
+        auto job = new LookupNameJob(this, GenTXID(), name, resultHandler);
```
Comment (tewinget) [inline llarp/service/endpoint.cpp:858]: ownership very unclear, I smell cursed shenanigans

File: llarp/service/endpoint.cpp (hunk 1 above)
Comment (majestrate) [reply llarp/service/endpoint.cpp]: fixed

File: llarp/service/endpoint.cpp (hunk 3 above)
Comment (majestrate) [reply llarp/service/endpoint.cpp:858]: beware the abyss

File: llarp/service/lns_tracker.hpp (hunk 2 above)
Comment (majestrate) [reply llarp/service/lns_tracker.hpp]: fixed

File: llarp/service/endpoint.cpp [hunk 4]
```diff
@@ -802,7 +802,7 @@ namespace llarp
     void
     Endpoint::LookupNameAsync(
         std::string name,
-        std::function<void(std::optional<std::variant<Address, RouterID>>)> handler)
+        std::function<void(std::optional<var::variant<Address, RouterID>>)> handler)
```
Comment (jagerman) [inline llarp/service/endpoint.cpp]: This should stay `std::variant`: the `var` function provides just `var::get` and `var::visit` which work on `std::variant` which mac *does* support as long as you don't use `std::get` or `std::visit` on them.

Comment (majestrate) [reply llarp/service/endpoint.cpp]: fixed

**Review by tewinget — APPROVED** [review]
should be fine, but still a bit ugly (re: handlers on handlers on handlers).

**Pull request merged**